Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,255 @@
+    internal class DefaultCredentialProvider
+    {
+        private static readonly ILogger Logger = ApplicationContext.Logger.ForType<DefaultCredentialProvider>();
+
+        /// <summary>Well known file which stores the default application credentials.</summary>
+        private const string WELL_KNOWN_CREDENTIALS_FILE = "application_default_credentials.json";
+
+        /// <summary>Environment variable override which stores the default application credentials file path.</summary>
+        private const string CREDENTIAL_ENV_VAR = "GOOGLE_APPLICATION_CREDENTIALS";
+
+        /// <summary>Environment variable which contains the Application Data settings.</summary>
+        private const string APPDATA_ENV_VAR = "APPDATA";
+
+        /// <summary>GCloud configuration directory in Windows, relative to %APPDATA%.</summary>
+        private const string CLOUDSDK_CONFIG_DIRECTORY = "gcloud";
+
+        /// <summary>Type name value of credentials file - user credentials.</summary>
+        private const string USER_FILE_TYPE = "authorized_user";
+
+        /// <summary>Type name value of credentials file - service account.</summary>
+        private const string SERVICE_ACCOUNT_FILE_TYPE = "service_account";
+
+        /// <summary>Help link to the application default credentials feature.</summary>
+        private const string HELP_PERMALINK = "https://developers.google.com/accounts/docs/application-default-credentials";
+
+        private object lockObject = new object();
+
+        // These variables should only be accessed inside a synchronized block.
+        private ICredential cachedCredential = null;
+        private bool checkedComputeEngine = false;
+
+        /// <summary>
+        /// Returns the Application Default Credentials which are credentials that identify and
+        /// authorize the whole application. This is the built-in service account if running on Google
+        /// Compute Engine or credentials specified by an environment variable or a file in a well-known location.
+        /// </summary>
+        internal ICredential GetApplicationDefaultCredential()
+        {
+            if (cachedCredential == null)
+            {
+                lock (lockObject)
+                {
+                    if (cachedCredential == null)
+                    {
+                        cachedCredential = GetDefaultCredentialUnsynchronized();
+                    }
+                }
+            }
+            if (cachedCredential != null)
+            {
+                return cachedCredential;
+            }
+
+            throw new InvalidOperationException(String.Format("The Application Default Credentials are not available. They are available if running"
+                + " in Google Compute Engine. Otherwise, the environment variable {0} must be defined"
+                + " pointing to a file defining the credentials. See {1} for more information.",
+                CREDENTIAL_ENV_VAR,
+                HELP_PERMALINK));
```

Comment:
As part of an earlier code review, I was asked to make this a const. I would leave it as it currently is, since there is no point in going back and forth.